OpenTelemetry Target Allocator

alexperez52 opened the pull request:
### Description

This PR addresses issue # 60 in the wg-prometheus repo. This is part 2 of a 2 part update in which enhancements are presented to the OTel Operator to support the deployment of a Target Allocator. This update contains the logic for the Operator Managed target allocator image which will be used to perform target sharding for the Prometheus scrape targets and expose them to the collector(s) for consumption using http sd config.

The Design document for this part 2 can be found here: 
https://drive.google.com/file/d/1K5-8KaZLKIf1EAq8552vCZPFHVWmiNWt/view?usp=sharing

### Type of change

New feature 

### Testing
Unit tests were added to test the least connection algorithm flow as well as updating and removing targets.
More testing details can be found on the doc mentioned above.

cc: @alolita @Aneurysm9 @rakyll @Raul9595

alexperez52 committed:
Least connection alogorithm logic added

Commit:
Added config parse and service discovery mechanisms

rakyll commented:
@Aneurysm9 I'm going to send some PRs for organization and styling. And it's easier to work on the otel-loadbalancer branch before reviewing this PR. Please don't review this PR yet.

rakyll committed:
Readability and clean up for discovery

* Accept all supported arguments when creating the discovery manager.
* Remove unused variables and redundant functions.
* Encapsulate things in Manager type for ease of use.
* Don't export TargetGroup, it's an internal type.
* Rename Get to ApplyConfig to make it clear it's updating the config.
* Rename Watch to Targets because Watch is not a watcher function.
* Add TODO to remove marshal/unmarshal.

rakyll committed:
Fix the TargetGroup

rakyll committed:
Config readability and cleanup

* Remove unused or redundant code.
* Use const instead of var for constants
* Move testing config files under config package testdata.
* Return ioutil.ReadFile error rather than masking it.
* Remove unneeded suite package.
* Add TODO to merge config package with the loadbalancer/mode.

Commit:
Merge pull request #8 from rakyll/discovery-clean

Readability and clean up for discovery

alexperez52 committed:
Merge pull request #9 from rakyll/config

Config readability and cleanup

rakyll committed:
Reorganize and cleanup the main

* Introduce a server type to make it easier to replace all the resources upon config change
* Make sure that the scheduler is cancelled before another one is started
* Make the main goroutine block on signals and watcher, start the server in a different goroutine.
* Make the signals and watchers select the main event loop.
* Add TODOs.

VineethReddy02 committed:
Bumps collector version to v0.30.0 (#342)

alexperez52 committed:
Merge pull request #10 from rakyll/main-program

Reorganize and cleanup the main

rakyll committed:
Cleanup mode package

* Add TODOs.
* Remove redundant types and change names.
* Unexport internal methods.

alexperez52 committed:
Merge pull request #12 from rakyll/lb

Cleanup mode package

rakyll committed:
Rename mode to sharder and more cleanup

* Unexport internal types.
* Rename variables.
* Move relevant methods together.
* Add TODOs.

Commit:
Update discovery.go to clean target extraction

Commit:
Merge pull request #14 from open-o11y/discover-update

Update discovery.go to clean target extraction

Commit:
Merge pull request #13 from rakyll/lb2

Rename mode to sharder and more cleanup

rakyll committed:
Fix the broken build by readding the TargetGroup

rakyll committed:
Fix the discovery deadlock

 <-m.manager.SyncCh() is going to return when there are new changes.
Targets() and ApplyConfig() were occasionally deadlocked when there
were no target changes. Introducing a watching mechanism to watch
the changes and report rather than periodically querying the targets.
This also helps us to eliminate the gocron dependency and main
program is much simpler now.

Also fixing and simplifying the tests.

rakyll committed:
Move HTTP handlers to the sharder

This makes the display cache more self contained in the sharder package for now.

dependabot[bot] committed:
Bump k8s.io/kubectl from 0.21.2 to 0.21.3 (#348)

Bumps [k8s.io/kubectl](https://github.com/kubernetes/kubectl) from 0.21.2 to 0.21.3.
- [Release notes](https://github.com/kubernetes/kubectl/releases)
- [Commits](https://github.com/kubernetes/kubectl/compare/v0.21.2...v0.21.3)

---
updated-dependencies:
- dependency-name: k8s.io/kubectl
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump docker/metadata-action from 3.4.0 to 3.4.1 (#344)

Bumps [docker/metadata-action](https://github.com/docker/metadata-action) from 3.4.0 to 3.4.1.
- [Release notes](https://github.com/docker/metadata-action/releases)
- [Commits](https://github.com/docker/metadata-action/compare/v3.4.0...v3.4.1)

---
updated-dependencies:
- dependency-name: docker/metadata-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

VineethReddy02 committed:
Bump kubectl-kuttl to v0.11.0 (#343)

Commit:
Merge pull request #15 from rakyll/discovery-nonblocking

Fix the discovery deadlock

alexperez52 committed:
Merge pull request #16 from rakyll/http

Move HTTP handlers to the sharder

rakyll committed:
Test discovery with a file_sd_configs

To make sure the discovery manager is handling the file_sd_configs.

Commit:
Merge pull request #18 from rakyll/tests

Test discovery with a file_sd_configs

Commit:
Updated main.go server

rsvarma95 committed:
Updated collector.go to reflect latest docker image and added Docker file

Commit:
Updated collector.go to handle new collector instances

Commit:
Update collector.go

Commit:
Update collector.go

Commit:
Added ctx check and loop break on timeout

Commit:
Merge pull request #19 from open-o11y/handle-collectors

Handle updated collector instances

alexperez52 committed:
Renamed from sharder to allocator & merged ttargetItem and targets

alexperez52 committed:
Merge error fixed

alexperez52 committed:
remove build file

alexperez52 committed:
Updated naming in main.go file

alexperez52 committed:
Added mutex locks and removed some TODO comments

Commit:
Testing suite cleanup

alexperez52 committed:
Renamed folder and respective package imports

alexperez52 committed:
Merge branch 'otel-loadbalancer' of https://github.com/open-o11y/opentelemetry-operator into otel-loadbalancer

Aneurysm9 commented on otel-allocator/allocation/allocator.go:
```diff
@@ -0,0 +1,130 @@
+func (allocator *Allocator) SetTargets(targets []TargetItem) {
+	// Dump old data
+	allocator.m.Lock()
+	for k := range allocator.targetsWaiting {
+		delete(allocator.targetsWaiting, k)
+	}
```
```suggestion
	allocator.targetsWaiting = make(map[string]TargetItem)
```

Don't think it's necessary to delete each entry individually, just replace the whole map.

Aneurysm9 commented on otel-allocator/allocation/allocator.go:
```diff
@@ -0,0 +1,130 @@
+		log.Fatal("no collector instances present")
+	}
+	allocator.m.Lock()
+	for _, i := range collectors {
```
Does the collector map also need to be cleared here?  What happens when a collector disappears?  I don't think this would remove it from the list.

alexperez52 replied:
For now, when there is a change with the collectors, we are doing a full re-balance. SetCollectors() will be called with a brand new list of collectors and these collectors are all guaranteed to be up.

Aneurysm9 commented on otel-loadbalancer/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+# Build the target allocator binary
```
Looks like this got duplicated.  Ensure to `git rm -r` the old directory.  Maybe also squash the commits on this branch before making a PR upstream.

Aneurysm9 reviewed (approved):
Couple minor notes.  Only thing that must be addressed before sending it upstream is the duplicate directory.

alexperez52 committed:
Removed duplicate folder

alexperez52 committed:
Changed delete iteration to just creating a new map

alexperez52 committed:
Removed TODOs that were complete

alexperez52 committed:
Minor change

alexperez52 committed:
Removed caching structure & change fsnotify to Write

alexperez52 committed:
Changed fsnotify back to create

rsvarma95 committed:
Minor changes

Commit:
Merge branch 'otel-loadbalancer' of https://github.com/open-o11y/opentelemetry-operator into otel-loadbalancer
